doc/watchexec.1
.ie \n(.g .ds Aq \(aq
.el .ds Aq '
.TH watchexec 1  "watchexec 2.8.1" 
.SH NAME
watchexec \- Execute commands when watched files change
.SH SYNOPSIS
\fBwatchexec\fR [\fB\-\-bell\fR] [\fB\-c\fR|\fB\-\-clear\fR] [\fB\-\-completions\fR] [\fB\-\-color\fR] [\fB\-d\fR|\fB\-\-debounce\fR] [\fB\-\-delay\-run\fR] [\fB\-e\fR|\fB\-\-exts\fR] [\fB\-E\fR|\fB\-\-env\fR] [\fB\-\-emit\-events\-to\fR] [\fB\-f\fR|\fB\-\-filter\fR] [\fB\-\-socket\fR] [\fB\-\-filter\-file\fR] [\fB\-j\fR|\fB\-\-filter\-prog\fR] [\fB\-\-fs\-events\fR] [\fB\-i\fR|\fB\-\-ignore\fR] [\fB\-I\fR|\fB\-\-interactive\fR] [\fB\-\-exit\-on\-error\fR] [\fB\-\-ignore\-file\fR] [\fB\-\-ignore\-nothing\fR] [\fB\-\-log\-file\fR] [\fB\-\-manual\fR] [\fB\-\-map\-signal\fR] [\fB\-n \fR] [\fB\-N\fR|\fB\-\-notify\fR] [\fB\-\-no\-default\-ignore\fR] [\fB\-\-no\-discover\-ignore\fR] [\fB\-\-no\-process\-group\fR] [\fB\-\-no\-global\-ignore\fR] [\fB\-\-no\-meta\fR] [\fB\-\-no\-project\-ignore\fR] [\fB\-\-no\-vcs\-ignore\fR] [\fB\-o\fR|\fB\-\-on\-busy\-update\fR] [\fB\-\-only\-emit\-events\fR] [\fB\-\-poll\fR] [\fB\-\-print\-events\fR] [\fB\-\-project\-origin\fR] [\fB\-p\fR|\fB\-\-postpone\fR] [\fB\-q\fR|\fB\-\-quiet\fR] [\fB\-Q\fR|\fB\-\-quote\fR] [\fB\-r\fR|\fB\-\-restart\fR] [\fB\-s\fR|\fB\-\-signal\fR] [\fB\-\-shell\fR] [\fB\-\-stdin\-quit\fR] [\fB\-\-stop\-signal\fR] [\fB\-\-stop\-timeout\fR] [\fB\-\-timeout\fR] [\fB\-\-timings\fR] [\fB\-v\fR|\fB\-\-verbose\fR]... [\fB\-w\fR|\fB\-\-watch\fR] [\fB\-\-workdir\fR] [\fB\-W\fR|\fB\-\-watch\-non\-recursive\fR] [\fB\-\-wrap\-process\fR] [\fB\-F\fR|\fB\-\-watch\-file\fR] [\fB\-\-no\-follow\-symlinks\fR] [\fB\-h\fR|\fB\-\-help\fR] [\fB\-V\fR|\fB\-\-version\fR] [\fICOMMAND\fR] 
.SH DESCRIPTION
Execute commands when watched files change.
.PP
Recursively monitors the current directory for changes, executing the command when a filesystem change is detected (among other event sources). By default, watchexec uses efficient kernel\-level mechanisms to watch for changes.
.PP
At startup, the specified command is run once, and watchexec begins monitoring for changes.
.PP
Events are debounced and checked using a variety of mechanisms, which you can control using the flags in the **Filtering** section. The order of execution is: internal prioritisation (signals come before everything else, and SIGINT/SIGTERM are processed even more urgently), then file event kind (`\-\-fs\-events`), then files explicitly watched with `\-w`, then ignores (`\-\-ignore` and co), then filters (which includes `\-\-exts`), then filter programs.
.PP
Examples:
.PP
Rebuild a project when source files change:
.PP
$ watchexec make
.PP
Watch all HTML, CSS, and JavaScript files for changes:
.PP
$ watchexec \-e html,css,js make
.PP
Run tests when source files change, clearing the screen each time:
.PP
$ watchexec \-c make test
.PP
Launch and restart a node.js server:
.PP
$ watchexec \-r node app.js
.PP
Watch lib and src directories for changes, rebuilding each time:
.PP
$ watchexec \-w lib \-w src make
.SH OPTIONS
.TP
\fB\-\-completions\fR \fI<SHELL>\fR
Generate a shell completions script

Provides a completions script or configuration for the given shell. If Watchexec is not distributed with pre\-generated completions, you can use this to generate them yourself.

Supported shells: bash, elvish, fish, nu, powershell, zsh.
.TP
\fB\-\-manual\fR
Show the manual page

This shows the manual page for Watchexec, if the output is a terminal and the \*(Aqman\*(Aq program is available. If not, the manual page is printed to stdout in ROFF format (suitable for writing to a watchexec.1 file).
.TP
\fB\-\-only\-emit\-events\fR
Only emit events to stdout, run no commands.

This is a convenience option for using Watchexec as a file watcher, without running any commands. It is almost equivalent to using `cat` as the command, except that it will not spawn a new process for each event.

This option implies `\-\-emit\-events\-to=json\-stdio`; you may also use the text mode by specifying `\-\-emit\-events\-to=stdio`.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help (see a summary with \*(Aq\-h\*(Aq)
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version
.TP
[\fICOMMAND\fR]
Command (program and arguments) to run on changes

It\*(Aqs run when events pass filters and the debounce period (and once at startup unless \*(Aq\-\-postpone\*(Aq is given). If you pass flags to the command, you should separate it with \-\- though that is not strictly required.

Examples:

$ watchexec \-w src npm run build

$ watchexec \-w src \-\- rsync \-a src dest

Take care when using globs or other shell expansions in the command. Your shell may expand them before ever passing them to Watchexec, and the results may not be what you expect. Compare:

$ watchexec echo src/*.rs

$ watchexec echo \*(Aqsrc/*.rs\*(Aq

$ watchexec \-\-shell=none echo \*(Aqsrc/*.rs\*(Aq

Behaviour depends on the value of \*(Aq\-\-shell\*(Aq: for all except \*(Aqnone\*(Aq, every part of the command is joined together into one string with a single ascii space character, and given to the shell as described in the help for \*(Aq\-\-shell\*(Aq. For \*(Aqnone\*(Aq, each distinct element the command is passed as per the execvp(3) convention: first argument is the program, as a path or searched for in the \*(AqPATH\*(Aq environment variable, rest are arguments.
.SH COMMAND
.TP
\fB\-\-delay\-run\fR \fI<DURATION>\fR
Sleep before running the command

This option will cause Watchexec to sleep for the specified amount of time before running the command, after an event is detected. This is like using "sleep 5 && command" in a shell, but portable and slightly more efficient.

Takes a unit\-less value in seconds, or a time span value such as "2min 5s". Providing a unit\-less value is deprecated and will warn; it will be an error in the future.
.TP
\fB\-E\fR, \fB\-\-env\fR \fI<KEY=VALUE>\fR
Add env vars to the command

This is a convenience option for setting environment variables for the command, without setting them for the Watchexec process itself.

Use key=value syntax. Multiple variables can be set by repeating the option.
.TP
\fB\-\-socket\fR \fI<PORT>\fR
Provide a socket to the command

This implements the systemd socket\-passing protocol, like with `systemfd`: sockets are opened from the watchexec process, and then passed to the commands it runs. This lets you keep sockets open and avoid address reuse issues or dropping packets.

This option can be supplied multiple times, to open multiple sockets.

The value can be either of `PORT` (opens a TCP listening socket at that port), `HOST:PORT` (specify a host IP address; IPv6 addresses can be specified `[bracketed]`), `TYPE::PORT` or `TYPE::HOST:PORT` (specify a socket type, `tcp` / `udp`).

This integration only provides basic support, if you want more control you should use the `systemfd` tool from <https://github.com/mitsuhiko/systemfd>, upon which this is based. The syntax here and the spawning behaviour is identical to `systemfd`, and both watchexec and systemfd are compatible implementations of the systemd socket\-activation protocol.

Watchexec does _not_ set the `LISTEN_PID` variable on unix, which means any child process of your command could accidentally bind to the sockets, unless the `LISTEN_*` variables are removed from the environment.
.TP
\fB\-n\fR
Shorthand for \*(Aq\-\-shell=none\*(Aq
.TP
\fB\-\-no\-process\-group\fR
Don\*(Aqt use a process group

By default, Watchexec will run the command in a process group, so that signals and terminations are sent to all processes in the group. Sometimes that\*(Aqs not what you want, and you can disable the behaviour with this option.

Deprecated, use \*(Aq\-\-wrap\-process=none\*(Aq instead.
.TP
\fB\-Q\fR, \fB\-\-quote\fR
Set whether or not to quote the command symbols when passing them to the shell.

On Windows by default this is treated as false, as CMD and PowerShell do not work correctly when the symbols are quoted. For git\-bash and nushell on Windows, as well as other shells that won\*(Aqt work correctly without quoting, opt in to quoting using this option.

On Linux and MacOS this is ignored and always treated as true, because they do not allow raw values to be passed.

The difference between the two operating systems is to allow both to work as expected by default for their built in shells, and to allow changing the default behavior for non\-default shells if necessary.

If you wish to write shell scripts that work on Linux/MacOS as well as Windows via git\-bash, you should set this option to true so that the behavior is the same on all platforms. The same goes for Nushell, and possibly other newer shells.
.TP
\fB\-\-shell\fR \fI<SHELL>\fR
Use a different shell

By default, Watchexec will use \*(Aq$SHELL\*(Aq if it\*(Aqs defined or a default of \*(Aqsh\*(Aq on Unix\-likes, and either \*(Aqpwsh\*(Aq, \*(Aqpowershell\*(Aq, or \*(Aqcmd\*(Aq (CMD.EXE) on Windows, depending on what Watchexec detects is the running shell.

With this option, you can override that and use a different shell, for example one with more features or one which has your custom aliases and functions.

If the value has spaces, it is parsed as a command line, and the first word used as the shell program, with the rest as arguments to the shell.

The command is run with the \*(Aq\-c\*(Aq flag (except for \*(Aqcmd\*(Aq on Windows, where it\*(Aqs \*(Aq/C\*(Aq).

The special value \*(Aqnone\*(Aq can be used to disable shell use entirely. In that case, the command provided to Watchexec will be parsed, with the first word being the executable and the rest being the arguments, and executed directly. Note that this parsing is rudimentary, and may not work as expected in all cases.

Using \*(Aqnone\*(Aq is a little more efficient and can enable a stricter interpretation of the input, but it also means that you can\*(Aqt use shell features like globbing, redirection, control flow, logic, or pipes.

Examples:

Use without shell:

$ watchexec \-n \-\- zsh \-x \-o shwordsplit scr

Use with powershell core:

$ watchexec \-\-shell=pwsh \-\- Test\-Connection localhost

Use with CMD.exe:

$ watchexec \-\-shell=cmd \-\- dir

Use with a different unix shell:

$ watchexec \-\-shell=bash \-\- \*(Aqecho $BASH_VERSION\*(Aq

Use with a unix shell and options:

$ watchexec \-\-shell=\*(Aqzsh \-x \-o shwordsplit\*(Aq \-\- scr
.TP
\fB\-\-stop\-signal\fR \fI<SIGNAL>\fR
Signal to send to stop the command

This is used by \*(Aqrestart\*(Aq and \*(Aqsignal\*(Aq modes of \*(Aq\-\-on\-busy\-update\*(Aq (unless \*(Aq\-\-signal\*(Aq is provided). The restart behaviour is to send the signal, wait for the command to exit, and if it hasn\*(Aqt exited after some time (see \*(Aq\-\-timeout\-stop\*(Aq), forcefully terminate it.

The default on unix is "SIGTERM".

Input is parsed as a full signal name (like "SIGTERM"), a short signal name (like "TERM"), or a signal number (like "15"). All input is case\-insensitive.

On Windows this option is technically supported but only supports the "KILL" event, as Watchexec cannot yet deliver other events. Windows doesn\*(Aqt have signals as such; instead it has termination (here called "KILL" or "STOP") and "CTRL+C", "CTRL+BREAK", and "CTRL+CLOSE" events. For portability the unix signals "SIGKILL", "SIGINT", "SIGTERM", and "SIGHUP" are respectively mapped to these.
.TP
\fB\-\-stop\-timeout\fR \fI<TIMEOUT>\fR
Time to wait for the command to exit gracefully

This is used by the \*(Aqrestart\*(Aq mode of \*(Aq\-\-on\-busy\-update\*(Aq. After the graceful stop signal is sent, Watchexec will wait for the command to exit. If it hasn\*(Aqt exited after this time, it is forcefully terminated.

Takes a unit\-less value in seconds, or a time span value such as "5min 20s". Providing a unit\-less value is deprecated and will warn; it will be an error in the future.

The default is 10 seconds. Set to 0 to immediately force\-kill the command.

This has no practical effect on Windows as the command is always forcefully terminated; see \*(Aq\-\-stop\-signal\*(Aq for why.
.TP
\fB\-\-timeout\fR \fI<TIMEOUT>\fR
Kill the command if it runs longer than this duration

Takes a time span value such as "30s", "5min", or "1h 30m".

When the timeout is reached, the command is gracefully stopped using \-\-stop\-signal, then forcefully terminated after \-\-stop\-timeout if still running.

Each run of the command has its own independent timeout.
.TP
\fB\-\-workdir\fR \fI<DIRECTORY>\fR
Set the working directory

By default, the working directory of the command is the working directory of Watchexec. You can change that with this option. Note that paths may be less intuitive to use with this.
.TP
\fB\-\-wrap\-process\fR \fI<MODE>\fR [default: auto]
Configure how the process is wrapped

By default (mode \*(Aqauto\*(Aq), Watchexec will run the command in a session on Mac, in a process group in Unix, and in a Job Object in Windows.

Some Unix programs prefer running in a session, while others do not work in a process group. Additionally, on Unix in \*(Aqauto\*(Aq mode, Watchexec will detect when a program needs to be the foreground process to handle terminal input, and grant that temporarily. This is the case for many pagers and interactive programs. To opt\-out, specify an explicit mode as needed.

Use \*(Aqgroup\*(Aq to use a process group, \*(Aqsession\*(Aq to use a process session, and \*(Aqnone\*(Aq to run the command directly. On Windows, either of \*(Aqgroup\*(Aq or \*(Aqsession\*(Aq will use a Job Object.

If you find you need to specify this frequently for different kinds of programs, file an issue at <https://github.com/watchexec/watchexec/issues>. As errors of this nature are hard to debug and can be highly environment\-dependent, reports from *multiple affected people* are more likely to be actioned promptly. Ask your friends/colleagues!
.SH EVENTS
.TP
\fB\-d\fR, \fB\-\-debounce\fR \fI<TIMEOUT>\fR
Time to wait for new events before taking action

When an event is received, Watchexec will wait for up to this amount of time before handling it (such as running the command). This is essential as what you might perceive as a single change may actually emit many events, and without this behaviour, Watchexec would run much too often. Additionally, it\*(Aqs not infrequent that file writes are not atomic, and each write may emit an event, so this is a good way to avoid running a command while a file is partially written.

An alternative use is to set a high value (like "30min" or longer), to save power or bandwidth on intensive tasks, like an ad\-hoc backup script. In those use cases, note that every accumulated event will build up in memory.

Takes a unit\-less value in milliseconds, or a time span value such as "5sec 20ms". Providing a unit\-less value is deprecated and will warn; it will be an error in the future.

The default is 50 milliseconds. Setting to 0 is highly discouraged.
.TP
\fB\-\-emit\-events\-to\fR \fI<MODE>\fR
Configure event emission

Watchexec can emit event information when running a command, which can be used by the child
process to target specific changed files.

One thing to take care with is assuming inherent behaviour where there is only chance.
Notably, it could appear as if the `RENAMED` variable contains both the original and the new
path being renamed. In previous versions, it would even appear on some platforms as if the
original always came before the new. However, none of this was true. It\*(Aqs impossible to
reliably and portably know which changed path is the old or new, "half" renames may appear
(only the original, only the new), "unknown" renames may appear (change was a rename, but
whether it was the old or new isn\*(Aqt known), rename events might split across two debouncing
boundaries, and so on.

This option controls where that information is emitted. It defaults to \*(Aqnone\*(Aq, which doesn\*(Aqt
emit event information at all. The other options are \*(Aqenvironment\*(Aq (deprecated), \*(Aqstdio\*(Aq,
\*(Aqfile\*(Aq, \*(Aqjson\-stdio\*(Aq, and \*(Aqjson\-file\*(Aq.

The \*(Aqstdio\*(Aq and \*(Aqfile\*(Aq modes are text\-based: \*(Aqstdio\*(Aq writes absolute paths to the stdin of
the command, one per line, each prefixed with `create:`, `remove:`, `rename:`, `modify:`,
or `other:`, then closes the handle; \*(Aqfile\*(Aq writes the same thing to a temporary file, and
its path is given with the $WATCHEXEC_EVENTS_FILE environment variable.

There are also two JSON modes, which are based on JSON objects and can represent the full
set of events Watchexec handles. Here\*(Aqs an example of a folder being created on Linux:

```json
  {
    "tags": [
      {
        "kind": "path",
        "absolute": "/home/user/your/new\-folder",
        "filetype": "dir"
      },
      {
        "kind": "fs",
        "simple": "create",
        "full": "Create(Folder)"
      },
      {
        "kind": "source",
        "source": "filesystem",
      }
    ],
    "metadata": {
      "notify\-backend": "inotify"
    }
  }
```

The fields are as follows:

  \- `tags`, structured event data.
  \- `tags[].kind`, which can be:
    * \*(Aqpath\*(Aq, along with:
      + `absolute`, an absolute path.
      + `filetype`, a file type if known (\*(Aqdir\*(Aq, \*(Aqfile\*(Aq, \*(Aqsymlink\*(Aq, \*(Aqother\*(Aq).
    * \*(Aqfs\*(Aq:
      + `simple`, the "simple" event type (\*(Aqaccess\*(Aq, \*(Aqcreate\*(Aq, \*(Aqmodify\*(Aq, \*(Aqremove\*(Aq, or \*(Aqother\*(Aq).
      + `full`, the "full" event type, which is too complex to fully describe here, but looks like \*(AqGeneral(Precise(Specific))\*(Aq.
    * \*(Aqsource\*(Aq, along with:
      + `source`, the source of the event (\*(Aqfilesystem\*(Aq, \*(Aqkeyboard\*(Aq, \*(Aqmouse\*(Aq, \*(Aqos\*(Aq, \*(Aqtime\*(Aq, \*(Aqinternal\*(Aq).
    * \*(Aqkeyboard\*(Aq, along with:
      + `keycode`. Currently only the value \*(Aqeof\*(Aq is supported.
    * \*(Aqprocess\*(Aq, for events caused by processes:
      + `pid`, the process ID.
    * \*(Aqsignal\*(Aq, for signals sent to Watchexec:
      + `signal`, the normalised signal name (\*(Aqhangup\*(Aq, \*(Aqinterrupt\*(Aq, \*(Aqquit\*(Aq, \*(Aqterminate\*(Aq, \*(Aquser1\*(Aq, \*(Aquser2\*(Aq).
    * \*(Aqcompletion\*(Aq, for when a command ends:
      + `disposition`, the exit disposition (\*(Aqsuccess\*(Aq, \*(Aqerror\*(Aq, \*(Aqsignal\*(Aq, \*(Aqstop\*(Aq, \*(Aqexception\*(Aq, \*(Aqcontinued\*(Aq).
      + `code`, the exit, signal, stop, or exception code.
  \- `metadata`, additional information about the event.

The \*(Aqjson\-stdio\*(Aq mode will emit JSON events to the standard input of the command, one per
line, then close stdin. The \*(Aqjson\-file\*(Aq mode will create a temporary file, write the
events to it, and provide the path to the file with the $WATCHEXEC_EVENTS_FILE
environment variable.

Finally, the \*(Aqenvironment\*(Aq mode was the default until 2.0. It sets environment variables
with the paths of the affected files, for filesystem events:

$WATCHEXEC_COMMON_PATH is set to the longest common path of all of the below variables,
and so should be prepended to each path to obtain the full/real path. Then:

  \- $WATCHEXEC_CREATED_PATH is set when files/folders were created
  \- $WATCHEXEC_REMOVED_PATH is set when files/folders were removed
  \- $WATCHEXEC_RENAMED_PATH is set when files/folders were renamed
  \- $WATCHEXEC_WRITTEN_PATH is set when files/folders were modified
  \- $WATCHEXEC_META_CHANGED_PATH is set when files/folders\*(Aq metadata were modified
  \- $WATCHEXEC_OTHERWISE_CHANGED_PATH is set for every other kind of pathed event

Multiple paths are separated by the system path separator, \*(Aq;\*(Aq on Windows and \*(Aq:\*(Aq on unix.
Within each variable, paths are deduplicated and sorted in binary order (i.e. neither
Unicode nor locale aware).

This is the legacy mode, is deprecated, and will be removed in the future. The environment
is a very restricted space, while also limited in what it can usefully represent. Large
numbers of files will either cause the environment to be truncated, or may error or crash
the process entirely. The $WATCHEXEC_COMMON_PATH is also unintuitive, as demonstrated by the
multiple confused queries that have landed in my inbox over the years.
.TP
\fB\-I\fR, \fB\-\-interactive\fR
Respond to keypresses to quit, restart, stop, or pause

In interactive mode, Watchexec listens for keypresses and responds to them. Currently supported keys are: \*(Aqr\*(Aq to restart the command, \*(Aqs\*(Aq to stop the running command, \*(Aqp\*(Aq to toggle pausing the watch, and \*(Aqq\*(Aq to quit. This requires a terminal (TTY) and puts stdin into raw mode, so the child process will not receive stdin input.
.TP
\fB\-\-exit\-on\-error\fR
Exit when the command has an error

By default, Watchexec will continue to watch and re\-run the command after the command exits, regardless of its exit status. With this option, it will instead exit when the command completes with any non\-success exit status.

This is useful when running Watchexec in a process manager or container, where you want the container to restart when the command fails rather than hang waiting for file changes.
.TP
\fB\-\-map\-signal\fR \fI<SIGNAL:SIGNAL>\fR
Translate signals from the OS to signals to send to the command

Takes a pair of signal names, separated by a colon, such as "TERM:INT" to map SIGTERM to SIGINT. The first signal is the one received by watchexec, and the second is the one sent to the command. The second can be omitted to discard the first signal, such as "TERM:" to not do anything on SIGTERM.

If SIGINT or SIGTERM are mapped, then they no longer quit Watchexec. Besides making it hard to quit Watchexec itself, this is useful to send pass a Ctrl\-C to the command without also terminating Watchexec and the underlying program with it, e.g. with "INT:INT".

This option can be specified multiple times to map multiple signals.

Signal syntax is case\-insensitive for short names (like "TERM", "USR2") and long names (like "SIGKILL", "SIGHUP"). Signal numbers are also supported (like "15", "31"). On Windows, the forms "STOP", "CTRL+C", and "CTRL+BREAK" are also supported to receive, but Watchexec cannot yet deliver other "signals" than a STOP.
.TP
\fB\-o\fR, \fB\-\-on\-busy\-update\fR \fI<MODE>\fR
What to do when receiving events while the command is running

Default is to \*(Aqdo\-nothing\*(Aq, which ignores events while the command is running, so that changes that occur due to the command are ignored, like compilation outputs. You can also use \*(Aqqueue\*(Aq which will run the command once again when the current run has finished if any events occur while it\*(Aqs running, or \*(Aqrestart\*(Aq, which terminates the running command and starts a new one. Finally, there\*(Aqs \*(Aqsignal\*(Aq, which only sends a signal; this can be useful with programs that can reload their configuration without a full restart.

The signal can be specified with the \*(Aq\-\-signal\*(Aq option.
.TP
\fB\-\-poll\fR [\fI<INTERVAL>\fR]
Poll for filesystem changes

By default, and where available, Watchexec uses the operating system\*(Aqs native file system watching capabilities. This option disables that and instead uses a polling mechanism, which is less efficient but can work around issues with some file systems (like network shares) or edge cases.

Optionally takes a unit\-less value in milliseconds, or a time span value such as "2s 500ms", to use as the polling interval. If not specified, the default is 30 seconds. Providing a unit\-less value is deprecated and will warn; it will be an error in the future.

Aliased as \*(Aq\-\-force\-poll\*(Aq.
.TP
\fB\-p\fR, \fB\-\-postpone\fR
Wait until first change before running command

By default, Watchexec will run the command once immediately. With this option, it will instead wait until an event is detected before running the command as normal.
.TP
\fB\-r\fR, \fB\-\-restart\fR
Restart the process if it\*(Aqs still running

This is a shorthand for \*(Aq\-\-on\-busy\-update=restart\*(Aq.
.TP
\fB\-s\fR, \fB\-\-signal\fR \fI<SIGNAL>\fR
Send a signal to the process when it\*(Aqs still running

Specify a signal to send to the process when it\*(Aqs still running. This implies \*(Aq\-\-on\-busy\-update=signal\*(Aq; otherwise the signal used when that mode is \*(Aqrestart\*(Aq is controlled by \*(Aq\-\-stop\-signal\*(Aq.


crates/lib/src/watchexec.rs
use std::{
	fmt,
	future::Future,
	sync::{Arc, OnceLock},
};

use async_priority_channel as priority;
use atomic_take::AtomicTake;
use futures::TryFutureExt;
use miette::Diagnostic;
use tokio::{
	spawn,
	sync::{mpsc, Notify},
	task::{JoinHandle, JoinSet},
};
use tracing::{debug, error, trace};
use watchexec_events::{Event, Priority};

use crate::{
	action::{self, ActionHandler},
	changeable::ChangeableFn,
	error::{CriticalError, RuntimeError},
	sources::{fs, keyboard, signal},
	Config,
};

/// The main watchexec runtime.
///
/// All this really does is tie the pieces together in one convenient interface.
///
/// It creates the correct channels, spawns every available event sources, the action worker, the
/// error hook, and provides an interface to change the runtime configuration during the runtime,
/// inject synthetic events, and wait for graceful shutdown.
pub struct Watchexec {
	/// The configuration of this Watchexec instance.
	///
	/// Configuration can be changed at any time using the provided methods on [`Config`].
	///
	/// Treat this field as readonly: replacing it with a different instance of `Config` will not do
	/// anything except potentially lose you access to the actual Watchexec config. In normal use
	/// you'll have obtained `Watchexec` behind an `Arc` so that won't be an issue.
	///
	/// # Examples
	///
	/// Change the action handler:
	///
	/// ```no_run
	/// # use watchexec::Watchexec;
	/// let wx = Watchexec::default();
	/// wx.config.on_action(|mut action| {
	///     if action.signals().next().is_some() {
	///         action.quit();
	///     }
	///
	///     action
	/// });
	/// ```
	///
	/// Set paths to be watched:
	///
	/// ```no_run
	/// # use watchexec::Watchexec;
	/// let wx = Watchexec::new(|mut action| {
	///     if action.signals().next().is_some() {
	///         action.quit();
	///     } else {
	///         for event in action.events.iter() {
	///             println!("{event:?}");
	///         }
	///     }
	///
	///     action
	/// }).unwrap();
	///
	/// wx.config.pathset(["."]);
	/// ```
	pub config: Arc<Config>,
	start_lock: Arc<Notify>,
	event_input: priority::Sender<Event, Priority>,
	handle: Arc<AtomicTake<JoinHandle<Result<(), CriticalError>>>>,
}

impl Default for Watchexec {
	/// Instantiate with default config.
	///
	/// Note that this will panic if the constructor errors.
	///
	/// Prefer calling `new()` instead.
	fn default() -> Self {
		Self::with_config(Default::default()).expect("Use Watchexec::new() to avoid this panic")
	}
}

impl fmt::Debug for Watchexec {
	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
		f.debug_struct("Watchexec").finish_non_exhaustive()
	}
}

impl Watchexec {
	/// Instantiates a new `Watchexec` runtime given an initial action handler.
	///
	/// Returns an [`Arc`] for convenience; use [`try_unwrap`][Arc::try_unwrap()] to get the value
	/// directly if needed, or use `new_with_config`.
	///
	/// Look at the [`Config`] documentation for more on the required action handler.
	/// Watchexec will subscribe to most signals sent to the process it runs in and send them, as
	/// [`Event`]s, to the action handler. At minimum, you should check for interrupt/ctrl-c events
	/// and call `action.quit()` in your handler, otherwise hitting ctrl-c will do nothing.
	pub fn new(
		action_handler: impl (Fn(ActionHandler) -> ActionHandler) + Send + Sync + 'static,
	) -> Result<Arc<Self>, CriticalError> {
		let config = Config::default();
		config.on_action(action_handler);
		Self::with_config(config).map(Arc::new)
	}

	/// Instantiates a new `Watchexec` runtime given an initial async action handler.
	///
	/// This is the same as [`new`](fn@Self::new) except the action handler is async.
	pub fn new_async(
		action_handler: impl (Fn(ActionHandler) -> Box<dyn Future<Output = ActionHandler> + Send + Sync>)
			+ Send
			+ Sync
			+ 'static,
	) -> Result<Arc<Self>, CriticalError> {
		let config = Config::default();
		config.on_action_async(action_handler);
		Self::with_config(config).map(Arc::new)
	}

	/// Instantiates a new `Watchexec` runtime with a config.
	///
	/// This is generally not needed: the config can be changed after instantiation (before and
	/// after _starting_ Watchexec with `main()`). The only time this should be used is to set the
	/// "unchangeable" configuration items for internal details like buffer sizes for queues, or to
	/// obtain Self unwrapped by an Arc like `new()` does.
	pub fn with_config(config: Config) -> Result<Self, CriticalError> {
		debug!(?config, pid=%std::process::id(), version=%env!("CARGO_PKG_VERSION"), "initialising");
		let config = Arc::new(config);
		let outer_config = config.clone();

		let notify = Arc::new(Notify::new());
		let start_lock = notify.clone();

		let (ev_s, ev_r) =
			priority::bounded(config.event_channel_size.try_into().unwrap_or(u64::MAX));
		let event_input = ev_s.clone();

		trace!("creating main task");
		let handle = spawn(async move {
			trace!("waiting for start lock");
			notify.notified().await;
			debug!("starting main task");

			let (er_s, er_r) = mpsc::channel(config.error_channel_size);

			let mut tasks = JoinSet::new();

			tasks.spawn(action::worker(config.clone(), er_s.clone(), ev_r).map_ok(|()| "action"));
			tasks.spawn(fs::worker(config.clone(), er_s.clone(), ev_s.clone()).map_ok(|()| "fs"));
			tasks.spawn(
				signal::worker(config.clone(), er_s.clone(), ev_s.clone()).map_ok(|()| "signal"),
			);
			tasks.spawn(
				keyboard::worker(config.clone(), er_s.clone(), ev_s.clone())
					.map_ok(|()| "keyboard"),
			);
			tasks.spawn(error_hook(er_r, config.error_handler.clone()).map_ok(|()| "error"));

			while let Some(Ok(res)) = tasks.join_next().await {
				match res {
					Ok("action") => {
						debug!("action worker exited, ending watchexec");
						break;
					}
					Ok(task) => {
						debug!(task, "worker exited");
					}
					Err(CriticalError::Exit) => {
						trace!("got graceful exit request via critical error, erasing the error");
						// Close event channel to signal worker task to stop
						ev_s.close();
					}
					Err(e) => {
						return Err(e);
					}
				}
			}

			debug!("main task graceful exit");
			tasks.shutdown().await;
			Ok(())
		});

		trace!("done with setup");
		Ok(Self {
			config: outer_config,
			start_lock,
			event_input,
			handle: Arc::new(AtomicTake::new(handle)),
		})
	}

	/// Inputs an [`Event`] directly.
	///
	/// This can be useful for testing, for custom event sources, or for one-off action triggers
	/// (for example, on start).
	///
	/// Hint: use [`Event::default()`] to send an empty event (which won't be filtered).
	pub async fn send_event(&self, event: Event, priority: Priority) -> Result<(), CriticalError> {
		self.event_input.send(event, priority).await?;
		Ok(())
	}

	/// Start watchexec and obtain the handle to its main task.
	///
	/// This must only be called once.
	///
	/// # Panics
	/// Panics if called twice.
	pub fn main(&self) -> JoinHandle<Result<(), CriticalError>> {
		trace!("notifying start lock");
		self.start_lock.notify_one();

		debug!("handing over main task handle");
		self.handle
			.take()
			.expect("Watchexec::main was called twice")
	}
}

async fn error_hook(
	mut errors: mpsc::Receiver<RuntimeError>,
	handler: ChangeableFn<ErrorHook, ()>,
) -> Result<(), CriticalError> {
	while let Some(err) = errors.recv().await {
		if matches!(err, RuntimeError::Exit) {
			trace!("got graceful exit request via runtime error, upgrading to crit");
			return Err(CriticalError::Exit);
		}

		error!(%err, "runtime error");
		let payload = ErrorHook::new(err);
		let crit = payload.critical.clone();
		handler.call(payload);
		ErrorHook::handle_crit(crit)?;
	}

	Ok(())
}

/// The environment given to the error handler.
///
/// This deliberately does not implement Clone to make it hard to move it out of the handler, which
/// you should not do.
///
/// The [`ErrorHook::critical()`] method should be used to send a [`CriticalError`], which will
/// terminate watchexec. This is useful to e.g. upgrade certain errors to be fatal.
///
/// Note that returning errors from the error handler does not result in critical errors.
#[derive(Debug)]
pub struct ErrorHook {
	/// The runtime error for which this handler was called.
	pub error: RuntimeError,
	critical: Arc<OnceLock<CriticalError>>,
}

impl ErrorHook {
	fn new(error: RuntimeError) -> Self {
		Self {
			error,
			critical: Default::default(),
		}
	}

	fn handle_crit(crit: Arc<OnceLock<CriticalError>>) -> Result<(), CriticalError> {
		match Arc::try_unwrap(crit) {
			Err(err) => {
				error!(?err, "error handler hook has an outstanding ref");
				Ok(())
			}
			Ok(crit) => crit.into_inner().map_or_else(
				|| Ok(()),
				|crit| {
					debug!(%crit, "error handler output a critical error");
					Err(crit)
				},
			),
		}
	}

	/// Set a critical error to be emitted.
	///
	/// This takes `self` and `ErrorHook` is not `Clone`, so it's only possible to call it once.
	/// Regardless, if you _do_ manage to call it twice, it will do nothing beyond the first call.
	pub fn critical(self, critical: CriticalError) {
		self.critical.set(critical).ok();
	}

	/// Elevate the current runtime error to critical.
	///
	/// This is a shorthand method for `ErrorHook::critical(CriticalError::Elevated(error))`.
	pub fn elevate(self) {
		let Self { error, critical } = self;
		critical
			.set(CriticalError::Elevated {
				help: error.help().map(|h| h.to_string()),
				err: error,
			})
			.ok();
	}
}
